pulsar-broker/src/main/java/org/apache/pulsar/compaction/CompactedTopicImpl.java

```diff
@@ -100,7 +100,13 @@ public void asyncReadEntriesOrWait(ManagedCursor cursor,
                                        boolean isFirstRead,
                                        ReadEntriesCallback callback, Consumer consumer) {
             PositionImpl cursorPosition;
-            if (isFirstRead && MessageId.earliest.equals(consumer.getStartMessageId())){
+            boolean readFromEarliest = false;
+            if (!cursor.isDurable() || ((ManagedCursorImpl) cursor).isCompactionCursor()
+                    || cursor.getMarkDeletedPosition() == null
+                    || cursor.getMarkDeletedPosition().getEntryId() == -1L) {
+                readFromEarliest = isFirstRead && MessageId.earliest.equals(consumer.getStartMessageId());
+            }
```

Comment:
You don't need to modify a deprecated method.

Reply:
This is for cherry-picking it to old release version.

Reply:
You'd better open a separated PR for cherry-pick.

Reply:
Yeah, I will open a separated PR for cherry-pick, but I think we can keep this change on the master.